### PR TITLE
Prepare Rust for opaque pointers

### PR DESCRIPTION
**Pull request by nikic**
Fix one codegen bug with opaque pointers, and update our IR tests to accept both typed pointer and opaque pointer IR. This is a bit annoying, but unavoidable if we want decent test coverage on both LLVM 14 and LLVM 15.

This prepares Rust for when LLVM will enable opaque pointers by default.

### CODE REVIEW

**Comment by eddyb** [comment]
@nikic Would it make sense to do another perf run with opaque pointers toggled off, but the same `llvm-project` submodule? To be able to compare across *just* "turning opaque pointers on" as opposed to any other changes to `llvm-project` potentially affecting the results.

File: src/test/codegen/abi-main-signature-32bit-c-int.rs [hunk 1]
```diff
@@ -7,4 +7,4 @@
 fn main() {
 }
 
-// CHECK: define i32 @main(i32{{( %0)?}}, i8**{{( %1)?}})
+// CHECK: define i32 @main(i32{{( %0)?}}, {{i8**|ptr}}{{( %1)?}})
```
Comment (eddyb) [inline src/test/codegen/abi-main-signature-32bit-c-int.rs:10]: This probably needs `i8\*\*` since it's inside a regex, right? (Similar thing in other files)

**Comment by nikic** [comment]
Testing baseline without opaque pointers now...

@bors try @rust-timer queue

**Comment by nikic** [comment]
> @nikic Would it make sense to do another perf run with opaque pointers toggled off, but the same `llvm-project` submodule? To be able to compare across _just_ "turning opaque pointers on" as opposed to any other changes to `llvm-project` potentially affecting the results.

Yeah, testing that now.

> **EDIT**: oh, wait, opaque pointers are coming entirely from the submodule switch? I guess in that case my idea would require adding a way to control it through FFI, sorry I missed that aspect at first.

It's possible to enable opaque-pointers per LLVMContext, but this currently doesn't play well with linker-plugin LTO at least (because opaque pointer bitcode will not automatically enable opaque pointer mode yet), so I'm patching the default on the LLVM side.

**Comment by nikic** [comment]
Comparison between the two: https://perf.rust-lang.org/compare.html?start=74beba27e6ec254300ee72d67e5be4d04031b50b&end=622e70f20deebb450e36b82cbabaa1a359e3b228&stat=instructions%3Au

**Comment by eddyb** [comment]
(Since this is tagged `[perf only]`, I marked it as draft, feel free to undo if anything changes)

**Comment by nikic** [comment]
Manual bisection identified the culprit as https://github.com/llvm/llvm-project/commit/5426da8ffa4a6d55adab21026ce6ebe8f1cc6ef2.

**Comment by nikic** [comment]
> Manual bisection identified the culprit as [llvm/llvm-project@5426da8](https://github.com/llvm/llvm-project/commit/5426da8ffa4a6d55adab21026ce6ebe8f1cc6ef2).

From some further debugging, `llvm-nm --print-armap` shows that the bitcode symbols are not present in the archive map. It looks like `CMAKE_RANLIB` is not passed and we end up using `ranlib` rather than `llvm-ranlib`, so that may be the issue. That said, running llvm-ranlib over the archive doesn't fix it either, but I'm not sure what ranlib is supposed to do for archives that already have an idnex.

**Comment by bjorn3** [comment]
Runtime performance seems to improve while compile times generally improve but sometimes regress.

File: src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs [hunk 2]
```diff
@@ -78,7 +78,7 @@ pub fn array_u8_1(f: fn(*const u8)) {
 #[no_mangle]
 pub fn array_u8_small(f: fn(*const u8)) {
     let a = [0u8; 2];
-    let b = [0u8; 8];
+    let b = [0u8; 7];
```
Comment (cuviper) [inline src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs:81]: Is this change somehow related to opaque pointers? Or is it something else incidental on llvm main? I don't see a problem with the change, but it seems curious.

**Review by cuviper — APPROVED** [review]
Whew, it will be a happy day when we can unify on `ptr` in all those tests...

File: src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs (hunk 2 above)
Comment (nikic) [reply src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs:81]: This is intended to test a heuristic that does not place a stack protector for arrays (strictly) smaller than 8, so it should be using size 7. The test still worked previously, because the `[8 x i8]` alloca was converted into an `i64` alloca, and did not receive the stack protector for that reason. This doesn't happen anymore with opaque pointers. The way these heuristics are currently implemented (inspecting alloca types in the backend) is unfortunately not stable under optimization (probably it should be a frontend responsibility to add metadata to allocas that should receive a protector, or something like that).

Comment (bjorn3) [reply src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs:81]: The documentation says that it is for any N > 8, not any N >= 8: https://github.com/rust-lang/rust/blob/ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491/compiler/rustc_codegen_llvm/src/lib.rs#L308

Comment (nikic) [reply src/test/assembly/stack-protector/stack-protector-heuristics-effect.rs:81]: Yeah, the documentation is wrong as well. I'll fix it. For reference, the actual logic is here: https://github.com/llvm/llvm-project/blob/65af17c7a00df3f8906cded34b63b309e687cf61/llvm/lib/CodeGen/StackProtector.cpp#L134
